### PR TITLE
Iac and Terratest tests for Secure AKS Construction Set

### PR DESCRIPTION
**Pull request by eedorenko**
# Iac and Terratest tests for Secure AKS Construction Set

## Description

- A set of terraform tests for Secure AKS Construction Set
- Iac Pipeline to deploy Secure AKS Construction Set with a Github Actions workflow  
- Iac Pipeline to deploy Secure AKS Construction Set with an Azure Pipeline
- Upgraded Flux to v2
- Included Flux V2 installation to IaC pipelines
- Azure Network Policy
- Upgrade AKS to 1.20.5
 
## Does this introduce a breaking change

- [ ] YES
- [X ] NO

### CODE REVIEW

File: .github/workflows/deploy-secure-aks-baseline.yaml [hunk 1]
```diff
@@ -0,0 +1,264 @@
+name: Deploy_Seccure_Aks_Baseline
```
Comment (dtzar) [inline .github/workflows/deploy-secure-aks-baseline.yaml]: Deploy_Seccure_Aks_Baseline --> Deploy_Secure_Aks_Baseline

File: enterprise_scale/construction_sets/aks/flux.tf [hunk 2]
```diff
@@ -0,0 +1,91 @@
+provider "flux" {}
+
+provider "kubectl" {
```
Comment (hieunhums) [inline enterprise_scale/construction_sets/aks/flux.tf]: I recently deprecate these kubeconfig values due to them being sensitive and us avoiding to store those on tfstate. #52 
Consider to use data source for aks to retrieve these kubeconfig data

Comment (hieunhums) [reply enterprise_scale/construction_sets/aks/flux.tf]: Flux.tf would fit better in addons folder

File: enterprise_scale/construction_sets/aks/flux_variables.tf [hunk 3]
```diff
@@ -0,0 +1,61 @@
+# Flux Variables
```
Comment (hieunhums) [inline enterprise_scale/construction_sets/aks/flux_variables.tf]: Consider to move to addons

File: enterprise_scale/construction_sets/aks/online/aks_secure_baseline/01-terraform.md [hunk 4]
```diff
@@ -52,18 +54,22 @@ cd caf-terraform-landingzones-starter/enterprise_scale/construction_sets/aks
 configuration_folder=online/aks_secure_baseline/configuration
 
 # Define the configuration files to apply, all tfvars files within the above folder recursively
```
Comment (hieunhums) [inline enterprise_scale/construction_sets/aks/online/aks_secure_baseline/01-terraform.md]: Define the configuration files to apply, all tfvars files within the above folder recursively except for launchpad subfolder which is not relevant for this standalone guide.

File: enterprise_scale/construction_sets/aks/online/aks_secure_baseline/01-terraform.md [hunk 5]
```diff
@@ -52,18 +54,22 @@ cd caf-terraform-landingzones-starter/enterprise_scale/construction_sets/aks
...
+
+# Define prefix for the resources 
+prefix=<MY_UNIQUE_PREFIX>
 
 # Load the CAF module and related providers
 terraform init -upgrade
 
 # Trigger the deployment of the resources
-eval terraform apply ${parameter_files}
+eval terraform apply ${parameter_files} -var test_prefix=$PREFIX
```
Comment (hieunhums) [inline enterprise_scale/construction_sets/aks/online/aks_secure_baseline/01-terraform.md]: Is test_prefix a custom var picked up from terratest? Is tag better suited?

File: enterprise_scale/construction_sets/aks/online/aks_secure_baseline/configuration/networking/firewall_application_rule_collection_definition.tfvars [hunk 6]
```diff
@@ -100,7 +100,10 @@ azurerm_firewall_application_rule_collection_definition = {
         ]
         target_fqdns = [
           "api.github.com",
+          "ghcr.io",
+          "*.ghcr.io",
           "github.com",
+          "*.githubusercontent.com",
```
Comment (hieunhums) [inline enterprise_scale/construction_sets/aks/online/aks_secure_baseline/configuration/networking/firewall_application_rule_collection_definition.tfvars]: let's consider if there is a way we can be more restrictive

File: enterprise_scale/construction_sets/aks/online/aks_secure_baseline/iac-pipeline.md [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+## Deploying levels manually
+Alternatively you can deploy the construction set level by level manually with *rover*.
+
+```bash
+# Go to the AKS construction set folder 
+cd caf-terraform-landingzones-starter
+
+# Start Rover container
+docker run -it --rm -v $(pwd):/tf/caf --user 0  aztfmod/rover:0.14.8-2103.1601 bash
```
Comment (hieunhums) [inline enterprise_scale/construction_sets/aks/online/aks_secure_baseline/iac-pipeline.md:67]: If opened in container in vscode then rover is available in bash. 
The env variables below will be supplied by the rover too after rover login or az login. 
I will help to refine this guide

File: enterprise_scale/construction_sets/aks/online/aks_secure_baseline/levels/1_foundation/parameters [hunk 8]
```diff
@@ -0,0 +1,5 @@
+global_settings.tfvars
```
Comment (hieunhums) [inline enterprise_scale/construction_sets/aks/online/aks_secure_baseline/levels/1_foundation/parameters:1]: Love this approach

**Review by hieunhums — CHANGES_REQUESTED** [review]
Thank you @eedorenko @sudivate for this great PR, love the content.
Please review some of the comments I have made. I have not fully gone through your PR yet but will resume.
Thanks again guys!

**Pull request merged**